8349944: [JMH] sun.misc.UnsafeOps cannot access class jdk.internal.misc.Unsafe

xyyNicole opened the pull request:
The UnsafeOps JMH benchmark fails with the following error:

      ```
      java.lang.IllegalAccessError: class org.openjdk.bench.sun.misc.UnsafeOps (in unnamed module @0x520a3426) cannot access class jdk.internal.misc.Unsafe (in module java.base) because module java.base does not export jdk.internal.misc to unnamed module @0x520a3426
      ```

Since this micro-benchmark is created for `sun.misc.Unsafe` rather than
`jdk.internal.misc.Unsafe`, we change it back as before JDK-8332744.

Note that even it will raise "proprietary API" warnings after this
patch, it is acceptable because the relevant APIs are bound for removal
for the integrity of the platform.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8349944](https://bugs.openjdk.org/browse/JDK-8349944): [JMH] sun.misc.UnsafeOps cannot access class jdk.internal.misc.Unsafe (**Bug** - P4)


### Reviewers
 * [Hao Sun](https://openjdk.org/census#haosun) (@shqking - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/23686/head:pull/23686` \
`$ git checkout pull/23686`

Update a local copy of the PR: \
`$ git checkout pull/23686` \
`$ git pull https://git.openjdk.org/jdk.git pull/23686/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 23686`

View PR using the GUI difftool: \
`$ git pr show -t 23686`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/23686.diff">https://git.openjdk.org/jdk/pull/23686.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/23686#issuecomment-2668054914)
</details>

xyyNicole commented:
/label add core-libs,hotspot-runtime

xyyNicole commented:
Hi @AlanBateman. Here is a bug fix for the micro benchmark of class Unsafe. Could you please review the changes? Thanks.

AlanBateman commented:
This micro was created for sun.misc.Unsafe, not jdk.internal.misc.Unsafe.  The change for JDK-8332744 should not have changed this test.

liach commented:
Yeah, a better fix for this would be to change the import of jdk.internal.misc back to sun.misc, as before JDK-8332744. Unfortunately we will have to accept the "proprietary API" warning, but they are necessary as such APIs are bound for removal for the integrity of the platform.

AlanBateman commented:
Thanks for restoring, this micro was specifically for sun.misc.Unsafe. I'm not wondering about the FFM micros that were also changed by JDK-8332744. Might not matter there but I think the original motive was to compare against sun.misc.Unsafe usage.

xyyNicole commented:
> Thanks for restoring, this micro was specifically for sun.misc.Unsafe. I'm not wondering about the FFM micros that were also changed by JDK-8332744. Might not matter there but I think the original motive was to compare against sun.misc.Unsafe usage.

Yeah. I have updated and only changed `Unsafe` back as you and @liach suggested, so that the runtime error here could be fixed.

xyyNicole commented:
Hi @cushon , I've prepared a rollback patch for your recent changes due to some balancing considerations. Could you please review and verify that? Thanks.

liach commented:
Hi @cushon, I believe disabling `JAVA_WARNINGS_ARE_ERRORS` should be done as part of those javac patches instead of part of this patch. We cannot verify such an addition in this patch, and it will just cause meaningless extra review and update cycles.

This flag is defined at https://github.com/openjdk/jdk/blob/964dd18fd2ba998e5c1efed48e15e516b0c22b19/make/common/JavaCompilation.gmk#L268. Unfortunately I don't know about make well enough to say what's the best way to disable this for micro build.

liach commented:
@AlanBateman Can you review this if this patch looks good in principle? Or should this bench be nuked like the virtual thread continuation one?

AlanBateman commented:
> @AlanBateman Can you review this if this patch looks good in principle? 

It needs confirmation that it doesn't break the build, I haven't looked into which parts of Liam's change weren't backed out. So if there are no build issues then it should be okay to integrate.

cushon commented:
> > @AlanBateman Can you review this if this patch looks good in principle?
> 
> It needs confirmation that it doesn't break the build, I haven't looked into which parts of Liam's change weren't backed out. So if there are no build issues then it should be okay to integrate.

All of the functional changes from JDK-8331081 and the redo got backed out, the only thing that stayed was test coverage, so I wouldn't expect this to break anything today.

Are the microbenchmarks built by CI? They weren't when JDK-8331081 originally happened, which was what led to it getting backed out the first time.

liach commented:
They are compiled by any test task, but they can still fail at run-time. That's why you see mostly fails-at-run fixes in the JDK.

shqking reviewed (approved):
>>>@AlanBateman Can you review this if this patch looks good in principle?
>>
>> It needs confirmation that it doesn't break the build, I haven't looked into which parts of Liam's change weren't backed out. So if there are no build issues then it should be okay to integrate.
>
> All of the functional changes from JDK-8331081 and the redo got backed out, the only thing that stayed was test coverage, so I wouldn't expect this to break anything today.

Yes. This PR doesn't break the build today as I tested locally.

As for the build failure as mentioned by @cushon , I can reproduce the build failure in my local environment.
1. how to reproduce the failure
It's important to note that we must revert the commit [1ab1c1d](https://github.com/openjdk/jdk/commit/1ab1c1d53b86228be85aac96fa5d69db39ac6317) firstly in order to reproduce the build failure.
```
the code: this PR with commit 1ab1c1d reverted
create the JMH bundle
configure with --with-jmh=path/to/the/jmh/jar
make images
make test-image # <--- we encounter build error here
```
2. possible way to mitigate the failure
As mentioned by @xyyNicole , new configure option `--disable-java-warnings-as-errors` is introduced recently in [7aeaa3c](https://github.com/openjdk/jdk/commit/7aeaa3c21c1420191fe8ff59e4cf99eae830754d). As I checked locally, the build failure is gone if we add this option.

```
the code: this PR with commit 1ab1c1d reverted
create the JMH bundle
configure with --with-jmh=path/to/the/jmh/jar --disable-java-warnings-as-errors
make images
make test-image # <--- no build error now
```

To be honest, I didn't fully understand [JDK-8349846](https://bugs.openjdk.org/browse/JDK-8349846). I was wondering if `--disable-java-warnings-as-errors` could help to address your concern about this PR. @cushon 

Here is my understanding for this PR:
1. As mentioned by @AlanBateman , "this micro was created for sun.misc.Unsafe, not jdk.internal.misc.Unsafe". We should keep this test intention as it was.
2. This PR doesn't break the build today.
3. As for the build error as mentioned by @cushon, it would be nice if `--disable-java-warnings-as-errors` can mitigate the error. If not, I agree with @liach that it's a different thing and out of the scope of this PR and we should handle it in a separate task.

cushon commented:
> I was wondering if `--disable-java-warnings-as-errors` could help to address your concern about this PR.

It doesn't address the issue unless it's configured by default, which I don't think is being proposed. I think the best approach would be to disable `-Werror` for specific compilations that are using `sun.misc` APIs, instead of disabling it for the entire build using `--disable-java-warnings-as-errors`. You can also proceed with this as-is and revisit if JDK-8349846 happens, I think it's up to you and the reviewers.

liach commented:
I have crafted a small build patch to allow benchmarks to build with all warnings not causing compilation failures. This is already the behavior for jtreg tests.

```
diff --git a/make/common/JavaCompilation.gmk b/make/common/JavaCompilation.gmk
index 70b3557baea..7e01e175406 100644
--- a/make/common/JavaCompilation.gmk
+++ b/make/common/JavaCompilation.gmk
@@ -271,7 +271,9 @@ define SetupJavaCompilationBody
   $1_FLAGS += $$($1_JAVAC_FLAGS)

   ifeq ($$(JAVA_WARNINGS_AS_ERRORS), true)
-    $1_FLAGS += -Werror
+    ifeq ($$($1_MUTE_ALL_JAVA_WARNINGS), )
+      $1_FLAGS += -Werror
+    endif
   endif

   ifneq ($$($1_DISABLED_WARNINGS), )
diff --git a/make/test/BuildMicrobenchmark.gmk b/make/test/BuildMicrobenchmark.gmk
index 92f40472c3c..d66b6a357c5 100644
--- a/make/test/BuildMicrobenchmark.gmk
+++ b/make/test/BuildMicrobenchmark.gmk
@@ -87,6 +87,7 @@ $(eval $(call SetupJavaCompilation, BUILD_JDK_MICROBENCHMARK, \
         serial preview dangling-doc-comments, \
     SRC := $(MICROBENCHMARK_SRC), \
     BIN := $(MICROBENCHMARK_CLASSES), \
+    MUTE_ALL_JAVA_WARNINGS := true, \
     JAVAC_FLAGS := \
         --add-exports java.base/jdk.internal.classfile.components=ALL-UNNAMED \
         --add-exports java.base/jdk.internal.classfile.impl=ALL-UNNAMED \
```

Feel free to take this patch and try building and running on latest mainline.

If you use this patch, please add me as a contributor and add this patch to the build list, like

> /contributor add liach
> /label add build

And make sure other engineers agree with this approach before integrating.
